oci/casext/gc.go

```diff
@@ -74,11 +77,25 @@ func (e Engine) GC(ctx context.Context) error {
 	}
 
 	n := 0
+sweep:
 	for _, digest := range blobs {
 		if _, ok := black[digest]; ok {
 			// Digest is in the black set.
 			continue
 		}
+
+		// Iterate over all policy funcs and if any one of them indicates that GC
+		// must not be performed, skip that blob
+		for _, policy := range policies {
+			ok, err := policy(digest)
```

Comment:
I wonder if this couldn't/shouldn't be a descriptor instead of a digest? I've written a lot of code switch()ing on mime types recently, so that would be handy...

Reply:
Had to change the signature of walk.go:Reachable(). Either that, or we need a different method in GC path.

Reply:
That's fine, this LGTM, thanks!

Reply:
You don't need to change the signature of `Reachable` (in fact, please don't do that) you can just use `Paths` (which gives you the set of `DescriptorPath`s you can reach -- and `DescriptorPaths` would be even better than `Descriptor`s). Maybe I should've called it `ReachablePaths`...

Given that this is all about adding policy functions, it does make sense to let those functions know what type the object is pointing to (though it is technically possible -- though unlikely -- that the same blob is referenced using different mime types).

Reply:
@cyphar Made the recommended changes. Would keep the policy funcs as agnostic about umoci's internals as possible, so would stay in the OCI domain.